dso-l1/src/main/java/com/tc/object/TransactionSource.java

```diff
@@ -19,32 +19,39 @@
 package com.tc.object;
 
 import com.tc.object.tx.TransactionID;
-
-import java.util.concurrent.ConcurrentSkipListSet;
+import java.util.BitSet;
 import java.util.concurrent.atomic.AtomicLong;
 
 public class TransactionSource {
 
-  private static final AtomicLong nextTransactionId = new AtomicLong(1L);
-
-  private static final ConcurrentSkipListSet<Long> pending = new ConcurrentSkipListSet<>();
+  private final AtomicLong current = new AtomicLong();
+  private long retiredBase = 1L;
+  private int retirePosition = 0;
+  private BitSet retired = new BitSet();
 
   public TransactionID create() {
-    long txn;
-    while (!pending.add((txn = nextTransactionId.getAndIncrement())));
-    return new TransactionID(txn);
+    return new TransactionID(current.incrementAndGet());
+  }
+
+  public synchronized TransactionID oldest() {
+    retirePosition = retired.nextClearBit(retirePosition);
+    long transactionId = retirePosition + retiredBase;
+    return new TransactionID(transactionId);
   }
 
-  public TransactionID oldest() {
-    Long first = pending.first();
-    if (first == null) {
-      return null;
-    } else {
-      return new TransactionID(first);
+  private void gc() {
+    if (retirePosition + retiredBase + (retirePosition>>8) > current.get()) {
```

Comment:
Not sure I follow the logic here. I would have expected the GC logic to be something like:
```java
if (retired.nextClearBit(0) > THRESHOLD) {
 //do gc
}

Reply:
it's a moving threshold.  The bigger the opportunity to reduce size the greater the chance of doing the work.  Trying to prevent gc while there are a large number of transactions in-flight